entmax/root_finding.py

```diff
@@ -133,6 +134,129 @@ def backward(cls, ctx, dY):
         return dX, None, None, None
 
 
+class NormmaxBisectFunction(Function):
+    @classmethod
+    def _gp(cls, x, alpha):
+        return x ** (alpha - 1)
+
+    @classmethod
+    def _gp_inv(cls, y, alpha):
+        return y ** (1 / (alpha - 1))
+
+    @classmethod
+    def _p(cls, X, alpha):
+        return cls._gp_inv(torch.clamp(X, min=0), alpha)
+
+    @classmethod
+    def forward(cls, ctx, X, alpha=2, dim=-1, n_iter=50):
+
+        if not isinstance(alpha, torch.Tensor):
+            alpha = torch.tensor(alpha, dtype=X.dtype, device=X.device)
+
+        alpha_shape = list(X.shape)
+        alpha_shape[dim] = 1
+        alpha = alpha.expand(*alpha_shape)
+
+        ctx.alpha = alpha
+        ctx.dim = dim
+        d = X.shape[dim]
+
+        max_val, _ = X.max(dim=dim, keepdim=True)
+        tau_lo = max_val - cls._gp(1, alpha)  # 1
+        tau_hi = max_val - cls._gp(1 / d, alpha)  # (1/d)**(alpha-1)
+
+        f_lo = (cls._p(X - tau_lo, alpha) ** alpha).sum(dim) - 1
```

Comment:
`f_lo` is unused. This is the same as on master, so this problem has been around for a while...